File: LeetCode-C++/LeetCode-C++/二叉树/_450_删除二叉搜索树中的节点.cpp
//
//  _450_删除二叉搜索树中的节点.cpp
//  LeetCode-C++
//
//  Created by warden on 2019/12/17.
//  Copyright © 2019 warden. All rights reserved.
//
// https://leetcode-cn.com/problems/delete-node-in-a-bst/

#include <stdio.h>
#include <vector>
#include "TreeNode.hpp"

using namespace std;

class Solution {
public:
    TreeNode* deleteNode(TreeNode* root, int key) {
        TreeNode* curNode = root;
        TreeNode* preNode = root;
        while (curNode && curNode->val != key) {
            preNode = curNode;
            if (curNode->val < key)
                curNode = curNode->right;
            else
                curNode = curNode->left;
        }
        if (!curNode) {
            return root;
        }
        if (!curNode->left && !curNode->right) {
            if (curNode == root) {
                return NULL;
            }
            if (preNode->left == curNode) {
                preNode->left = NULL;
            } else {
                preNode->right = NULL;
            }
        } else if (curNode->left && !curNode->right) {
            if (curNode == root) {
                return curNode->left;
            } else if (preNode->left == curNode) {
                preNode->left = curNode->left;
            } else {
                preNode->right = curNode->left;
            }
        } else if (curNode->right && !curNode->left) {
            if (curNode == root) {
                return curNode->right;
            } else if (preNode->left == curNode) {
                preNode->left = curNode->right;
            } else {
                preNode->right = curNode->right;
            }
        } else {
            TreeNode* dNode = curNode->left;
            while (dNode->right) {
                dNode = dNode->right;
            }
            dNode->right = curNode->right->left;
            curNode->right->left = curNode->left;
            if (curNode == root) {
                return curNode->right;
            } else if (preNode->left == curNode) {
                preNode->left = curNode->right;
            } else {
                preNode->right = curNode->right;
            }
        }
        return root;
    }
};

/**
 root = [5,3,6,2,4,null,7]
 key = 3

     5
    / \
   3   6
  / \   \
 2   4   7

 给定需要删除的节点值是 3，所以我们首先找到 3 这个节点，然后删除它。

 一个正确的答案是 [5,4,6,2,null,null,7], 如下图所示。

     5
    / \
   4   6
  /     \
 2       7

 另一个正确答案是 [5,2,6,null,4,null,7]。

     5
    / \
   2   6
    \   \
     4   7
 */
/**
int main(int argc, const char * argv[]) {
    vector<int> nodes = {5,3,6,2,4,INT_MAX,7};
    TreeNode* root = generateTreeByLevel(nodes);
    Solution* solution = new Solution();
    root = solution->deleteNode(root, 3);
    printTreeByLevel(root);
    return 0;
}
*/

File: LeetCode-Swift/LeetCode-Swift/二叉树/_450_删除二叉搜索树中的节点.swift
//
//  _450_删除二叉搜索树中的节点.swift
//  LeetCode-Swift
//
//  Created by warden on 2019/12/18.
//  Copyright © 2019 warden. All rights reserved.
//
// https://leetcode-cn.com/problems/delete-node-in-a-bst/

import Foundation

class _450_删除二叉搜索树中的节点 {
    func deleteNode(_ root: TreeNode?, _ key: Int) -> TreeNode? {
        var fatherNode: TreeNode?
        let deleteNode = findDeleteNode(root, key, &fatherNode)
        if let result = removeNode(deleteNode, fatherNode) {
            return result
        }
        if nil != root && key == root!.val  {
            return nil
        }
        return root
    }
        
    func findDeleteNode(_ root: TreeNode?, _ key: Int, _ father: inout TreeNode?) -> TreeNode? {
        guard root != nil, root?.val != key else { return root }
        var node = root
        while node != nil && node?.val != key {
            father = node
            if node!.val < key {
                node = node?.right
            } else {
                node = node?.left
            }
        }
        if node == nil {
            father = nil
        }
        return node
    }

    /// 删除节点，如果删除的是根节点，则返回新的根节点
    /// - Parameters:
    ///   - node: 要删除的节点
    ///   - father: 被删除节点的父节点
    func removeNode(_ node: TreeNode?, _ father: TreeNode?) -> TreeNode? {
        guard node != nil else { return nil }
        
        var child: TreeNode?    // 新的子节点，替代node位置的新节点
        if node?.left == nil || node?.right == nil {
            // 只有1个子节点或叶子节点
            child = node?.left ?? node?.right
            if father?.left === node {
                father?.left = child
            } else {
                father?.right = child
            }
            return father == nil ? child : nil
        } else {
            // 将右子节点作为新的子节点，并将原左子节点嫁接给右子节点当左子节点
            var dNode = node?.left
            child = node?.right
            while dNode?.right != nil {
                dNode = dNode?.right
            }
            dNode?.right = child?.left
            child?.left = node?.left
            if father?.left === node {
                father?.left = child
            } else {
                father?.right = child
            }
        }
        
        return father == nil ? child : nil
    }
}
